main/src/org/jetbrains/bsp/bazel/BazelBspServer.java

```diff
@@ -111,88 +109,46 @@
-  }
+  private final ProcessResolver processResolver;
+  private final QueryResolver queryResolver;
+  private final TargetsResolver targetsResolver;
+  private final ActionGraphResolver actionGraphResolver;
 
-  private boolean isInitialized() {
-    try {
-      isInitialized.get(1, TimeUnit.SECONDS);
-    } catch (InterruptedException | ExecutionException | TimeoutException e) {
-      return false;
-    }
-    return true;
-  }
-
-  private boolean isFinished() {
-    return isFinished.isDone();
-  }
+  private final ScalaBspServer scalaBspServer;
+  private final JavaBspServer javaBspServer;
 
-  private <T> CompletableFuture<T> completeExceptionally(ResponseError error) {
-    CompletableFuture<T> future = new CompletableFuture<>();
-    future.completeExceptionally(new ResponseErrorException(error));
-    return future;
-  }
+  // TODO: created in setter `setBuildClient`, HAVE TO BE moved to the constructor
+  private BuildClientLogger buildClientLogger;
 
-  private <T> CompletableFuture<T> handleBuildInitialize(
-      Supplier<Either<ResponseError, T>> request) {
-    if (isFinished())
-      return completeExceptionally(
-          new ResponseError(
-              ResponseErrorCode.serverErrorEnd, "Server has already shutdown!", false));
-    return getValue(request);
-  }
+  // TODO: imho bsp server creation on the server side is too ambiguous (constructor + setters)
+  public BazelBspServer(String pathToBazel) {
+    this.bazel = pathToBazel;
+    this.processResolver = new ProcessResolver(processLock, bazel, BES_BACKEND, PUBLISH_ALL_ACTIONS);
+    this.queryResolver = new QueryResolver(processResolver, processLock);
+    this.targetsResolver = new TargetsResolver(queryResolver);
+    this.actionGraphResolver = new ActionGraphResolver(processResolver);
 
-  private <T> CompletableFuture<T> handleBuildShutdown(Supplier<Either<ResponseError, T>> request) {
-    if (!isInitialized())
-      return completeExceptionally(
-          new ResponseError(
-              ResponseErrorCode.serverErrorEnd, "Server has not been initialized yet!", false));
-    return getValue(request);
+    this.scalaBspServer = new ScalaBspServer(targetsResolver, actionGraphResolver, SCALAC, JAVAC, getExecRoot());
+    this.javaBspServer = new JavaBspServer(targetsResolver, actionGraphResolver, JAVAC, KOTLINC, getExecRoot());
   }
 
-  private <T> CompletableFuture<T> executeCommand(Supplier<Either<ResponseError, T>> request) {
-    if (!isInitialized())
-      return completeExceptionally(
-          new ResponseError(
-              ResponseErrorCode.serverNotInitialized,
-              "Server has not been initialized yet!",
-              false));
-    if (isFinished())
-      return completeExceptionally(
-          new ResponseError(
-              ResponseErrorCode.serverErrorEnd, "Server has already shutdown!", false));
-
-    return getValue(request);
+  public void setBackendPort(int port) {
+    this.BES_BACKEND += port;
```

Comment:
BES_BACKEND looks like a constant, but it's not. And what if you call setBackendPort twice?

Reply:
I didn't change the logic, I only moved this method. Nevertheless I agree, it should be changed.